beps/0009-plugin-metadata/README.md

```diff
@@ -0,0 +1,173 @@
+
+In the case of new fields we always define them to be part of the top-level `"backstage"` field. The assumption is that any field that we define will always be tied to the Backstage ecosystem, even in the case of general fields such as "lifecycle". The only exception to this is if we believe that a field is generic enough that it could genuinely become part of the broader NPM ecosystem.
+
+All fields must be documented in a new section of the "Tooling" documentation on backstage.io. This documentation should include a description of the field, what it is used for, and what the expected values are. The documentation must also mention whether the field is expected to be filled in manually, or if it will be generated by tooling. Changes to this documentation must be approved by the Backstage core maintainers, it is not owned by individual project area maintainers.
+
+### Package Relationships
+
+This section defines how we associate packages from the same plugin with each other.
+
+Packages that are part of the same plugin should always be managed within the same monorepo and workspace. This does not apply to modules for a plugin, which may be hosted separately, but may still need to refer to the plugin that it is a module for.
+
+Each plugin package must define a `backstage.pluginId` field, which is the same identifier as is used in the implementation of the plugin. This field is inferred from the package name by the `backstage-cli repo fix` command if it is not present. It should only be defined for plugin package, for example `@backstage/errors` should not define a plugin ID. The `backstage.pluginId` field is required when publishing a package with a plugin or module role, or a library role with "plugin" in its name.
+
+The package relationships are defined in the `backstage.pluginPackages` field. The value of the field is an object where each key is the role of the package as defined by the `backstage.role` field, and the value is the package name. For example:
+
+```json
+{
+  "name": "@backstage/plugin-catalog",
+  "backstage": {
+    "role": "frontend-plugin",
+    "pluginId": "catalog",
+    "pluginPackages": {
+      "frontend-plugin": "@backstage/plugin-catalog",
+      "backend-plugin": "@backstage/plugin-catalog-backend",
+      "web-library": "@backstage/plugin-catalog-react",
+      "node-library": "@backstage/plugin-catalog-node",
+      "common-library": "@backstage/plugin-catalog-common"
```

Comment:
So the value of this is mainly that there's no effective way of doing the inverse lookup.

Still wondering if there's some point where we'll regret making this a single valued thing per role :) But sure, at that point maybe we just extend the format.

Still, I wonder if there's more metadata per entry that may be desired at some point. Like `optional: true` or whatever. Or where we break out parts of the core of a plugin into its own separate modules for structural reasons (catalog model implementation?) and ends up being a bit awkward to express here.

No big worries or action points - just thinking out loud.

Reply:
Another option here is to simply go with `"libraries": ["@backstage/plugin-catalog-react", "@backstage/plugin-catalog-node"]`, it doesn't give us as much info upfront, but maybe that's all redundant anyway?